### PR TITLE
Fix dangling volume vsphere detaches

### PR DESCRIPTION
**Pull request by gnufied**
Backports following PRs:

1. https://github.com/kubernetes/kubernetes/pull/96224 
2. https://github.com/kubernetes/kubernetes/pull/96689
3. https://github.com/kubernetes/kubernetes/pull/104410

/sig storage
/kind bug

```release-note
Fix detach from non-existant nodes and support dangling volumes
```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_util.go [hunk 1]
```diff
@@ -599,11 +601,131 @@ func (vs *VSphere) checkDiskAttached(ctx context.Context, nodes []k8stypes.NodeN
...
 	}
 	return nodesToRetry, nil
 }
 
+// BuildMissingVolumeNodeMap builds a map of volumes and nodes which are not known to attach detach controller.
+// There could be nodes in cluster which do not have any pods with vsphere volumes running on them
+// such nodes won't be part of disk verification check because attach-detach controller does not keep track
+// such nodes. But such nodes may still have dangling volumes on them and hence we need to scan all the
+// remaining nodes which weren't scanned by code previously.
+func (vs *VSphere) BuildMissingVolumeNodeMap(ctx context.Context) {
```
Comment (cheftako) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_util.go:617]: Does this need to be public?

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/nodemanager.go [hunk 2]
```diff
@@ -311,6 +311,17 @@ func (nm *NodeManager) GetNodeDetails() ([]NodeDetails, error) {
 	return nodeDetails, nil
 }
 
+// GetNodeNames returns list of nodes that are known to vsphere cloudprovider.
+// These are typically nodes that make up k8s cluster.
+func (nm *NodeManager) GetNodeNames() []k8stypes.NodeName {
```
Comment (cheftako) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/nodemanager.go:316]: Does this need to be public?

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_volume_map.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+func NewVsphereVolumeMap() *VsphereVolumeMap {
+	return &VsphereVolumeMap{
+		volumeNodeMap: map[volumePath]nodeVolumeStatus{},
+		nodeMap:       map[k8stypes.NodeName]bool{},
+	}
+}
+
+// StartDiskVerification marks all known volumes as unverified so as
+// disks which aren't verified can be removed at the end of verification process
+func (vsphereVolume *VsphereVolumeMap) StartDiskVerification() {
```
Comment (cheftako) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_volume_map.go:52]: Public?

**Comment by cheftako** [comment]
Missed this was a cherry pick.

**Comment by gnufied** [comment]
cc @kubernetes/release-managers

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**